### PR TITLE
User deepak topic

### PR DESCRIPTION
**Pull request by tavisca-dshrivastav**
@tavisca-sgakhar  please review

### CODE REVIEW

File: src/Tavisca.TMS.Core/Services/TopicService.cs [hunk 1]
```diff
@@ -18,14 +20,30 @@ public TopicService(ITopicRepository repository, IValidator<Topic> validator)
...
         }
 
+        public ApiResponse DeleteById(string id)
+        {
+      
+            if (!_validator.Validate(new Topic() { Id = id }, "Id").IsValid)
+                return new ApiFailureResponse(ErrorCodes.TopicApiError, new List<ErrorInfo> {new ErrorInfo(ErrorCodes.InvalidTopicId, ErrorMessages.InvalidTopicId) });
+            if (!_repository.Exists(id))
+                return new ApiFailureResponse(ErrorCodes.TopicApiError, new List<ErrorInfo> { new ErrorInfo(ErrorCodes.TopicIdNotFound, ErrorMessages.TopicIdNotFound) });
+            if (!_repository.Delete(id))
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Services/TopicService.cs]: Please make it more readable. It is fine if you use a variable or so for that.

File: src/Tavisca.TMS.Core/Services/TopicService.cs [hunk 2]
```diff
@@ -18,14 +20,30 @@ public TopicService(ITopicRepository repository, IValidator<Topic> validator)
...
+            else
+                return new ApiSuccessResponse(new { Message = "Topic deleted" });
+        }
+
+
+
         public ApiResponse Insert(Topic topic)
         {
+
+            topic.Id = Guid.NewGuid().ToString();
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Services/TopicService.cs]: what of topic is null?

File: src/Tavisca.TMS.Core/Validators/TopicValidator.cs [hunk 3]
```diff
@@ -7,13 +7,13 @@
...
 {
-    public class TopicValidator : AbstractValidator<Topic>
-    {
+    public class TopicValidator : AbstractValidator<Topic>    {
         private static Regex _specialCharacterPatternRegex = new Regex("^[a-zA-Z0-9]*$");
 
         public TopicValidator()
         {
-            //RuleFor(team => team.Id).Cascade(CascadeMode.StopOnFirstFailure).NotNull();
+            RuleFor(team => team.Id).NotNull().NotEmpty().WithMessage(ErrorMessages.InvalidTopicId)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Validators/TopicValidator.cs]: NotEmpty or Notnull why need both?

File: src/Tavisca.TMS.Core/Validators/TopicValidator.cs [hunk 4]
```diff
@@ -45,13 +45,23 @@ public TopicValidator()
                 .NotEmpty().WithMessage(ErrorMessages.TopicDurationIsBlank)
                 .Must(IsValidName).WithMessage(ErrorMessages.ContainsSpecialCharacters);
 
+           
+
+
             //Sub - Topic
             /*RuleFor(topic => topic.SubTopicId).Cascade(CascadeMode.StopOnFirstFailure)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Core/Validators/TopicValidator.cs]: Remove the commented code.

File: src/Tavisca.TMS.Persistence/Dal/TopicDal.cs [hunk 5]
```diff
@@ -52,5 +52,68 @@ public bool Insert(TopicDto topic)
...
+                    command.Parameters.AddWithValue("@topic_id", id);
+                    var topicCount = command.ExecuteNonQuery();
+                    command.Dispose();
+                    if (topicCount == 0)
+                        return false;
+                    return true;    
+                }
+                catch (Exception e)
+                {
+                    return false;
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Persistence/Dal/TopicDal.cs]: why are we suppressing the exception

File: src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs [hunk 6]
```diff
@@ -160,7 +161,50 @@ public void Test_for_validate_data_after_Insert_method_call_from_repository()
             });
             Assert.True(t2.GetType() == typeof(ApiFailureResponse));
         }
+        [Fact]
+        public void Should_return_api_failure_response_when_id_contains_whitespaces()
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs]: why we need fluentAssertions? does xUnit lacks certain functionalities which we want?

Comment (ckulkarni007) [inline src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs]: Unit test name should start from MethodName to indicate the method name

File: src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs [hunk 7]
```diff
@@ -160,7 +161,50 @@ public void Test_for_validate_data_after_Insert_method_call_from_repository()
...
+            _topicRepository.Setup(x => x.Delete("abc")).Returns(false);
+            _service.DeleteById("abc").Should().BeOfType<ApiFailureResponse>();
+        }
+        [Fact]
+        public void Should_return_api_success_response_when_id_is_valid_and_exists()
+        {
+            _service = new TopicService(_topicRepository.Object, new TopicValidator());
+            _topicRepository.Setup(x => x.Exists("abc")).Returns(true);
+            _topicRepository.Setup(x => x.Delete("abc")).Returns(true);
+            _service.DeleteById("abc").Should().BeOfType<ApiSuccessResponse>();
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs]: Only type check is not sufficient. You need to check the response parameters also.

File: src/Tavisca.TMS.Tests/Data/Dal/TopicDalFixture.cs [hunk 8]
```diff
@@ -34,5 +34,50 @@ public void When_null_connection_provided_it_should_return_null()
...
+        public void Should_return_false_when_topic_is_not_deleted_or_an_exception_occur(string id)
+        {
+            _mySqlConnection.Setup(x => x.GetConnection()).Returns(new MySqlConnectionProvider().GetConnection());
+            topicDal.DeleteById(id).Should().BeFalse();
+        }
+
+
+        [Theory]
+        [InlineData("3")]
+        public void Once_topic_deleted_topic_should_not_exist(string id)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Tests/Data/Dal/TopicDalFixture.cs]: Where is a unit test for a scenarios when any exception occurs? As per your code, it returns false, you need to validate it.

File: src/Tavisca.TMS.Web/Controllers/TopicController.cs [hunk 9]
```diff
@@ -28,5 +28,15 @@ public IActionResult Post([FromBody]Topic topic)
                 return BadRequest(result);
             return Ok(result);
         }
+
+        [HttpDelete("{id}")]
+        public IActionResult Delete(string id)
```
Comment (ckulkarni007) [inline src/Tavisca.TMS.Web/Controllers/TopicController.cs]: why this API is not asynchronous?

Comment (tavisca-dshrivastav) [reply src/Tavisca.TMS.Web/Controllers/TopicController.cs]: done

File: src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs (hunk 6 above)
Comment (tavisca-dshrivastav) [reply src/Tavisca.TMS.Tests/Core/Services/TopicServiceFixture.cs]: done

File: src/Tavisca.TMS.Core/Validators/TopicValidator.cs (hunk 3 above)
Comment (tavisca-dshrivastav) [reply src/Tavisca.TMS.Core/Validators/TopicValidator.cs]: done

File: src/Tavisca.TMS.Core/Services/TopicService.cs (hunk 1 above)
Comment (tavisca-dshrivastav) [reply src/Tavisca.TMS.Core/Services/TopicService.cs]: Done

**Pull request merged**